doc/developer-notes.md

```diff
@@ -160,6 +160,9 @@ A complete list of `@xxx` commands can be found at http://www.doxygen.nl/manual/
 As Doxygen recognizes the comments by the delimiters (`/**` and `*/` in this case), you don't
 *need* to provide any commands for a comment to be valid; just a description text is fine.
 
+Note: do not add todo or @todo entries in the source code. If something needs to be done, open an
+Issue instead.
```

Comment:
s/Issue/issue

Reply:
I think a rule without rationale is hard to understand. Maybe give some context and motivation?

Reply:
Also, it seems to strict and not allow for exceptions. For example if something is out-of-scope for a pull request, but still should be done in the long term, a TODO seems appropriate. Especially when the TODO is clearly motivated and clearly explains how to fix it.

For example:

* `boost::thread a; // TODO switch to C++11 std::thread, when possible`
* `GetCounter(); // TODO add lock annotation when class has been updated to our annotated mutex` (e.g. https://github.com/bitcoin/bitcoin/pull/17443#discussion_r425656746)
* ...

The advantage of inline TODO is that when the code gets removed or the TODO gets fixed, it will go away in the same commit. If it was filed as an issue, we'd have to walk the issue list and keep it in sync.

For general, non actionable TODOs, an issue makes sense. For example "increase encapsulation" or "increase performance with cache".

Thoughts?

Reply:
@MarcoFalke heh, you used my review as an example of a good TODO. I guess I should review less and start tackling TODOs :)

Reply:
Even "switch to std::thread" and "add lock annotations" can very much be issues (an umbrella issue could list the places).

My point is that these kind of long-term goals can change. Maybe lock annotations turn out to be more trouble than they're worth. Maybe std::thread would have turned out to have some problematic behavior on some platform. 
(it's not the case, these are fine in retrospect, but an issue allows updating these kind of things without touching the code)

But sure, it's possible to think of exceptions, but I think the general rule should be to not add TODOs in the source code.

Reply:
@MarcoFalke Both of those examples serve their purpose better if they are issues, instead of code comments, IMO, but I guess in certain cases (e.g. for things that must wait for something else to happen first) it's OK. Amended.

Reply:
> But sure, it's possible to think of exceptions, but I think the general rule should be to not add TODOs in the source code.

Agree. Concept ACK to have a policy so that TODOs are generally avoided.